Neos.Neos/Classes/Fusion/Cache/NeosFusionContextSerializer.php

```
@@ -0,0 +1,130 @@
+    {
+        if ($type === Node::class) {
+            return $this->deserializeNode($data);
+        }
+        return $this->fusionContextSerializer->denormalize($data, $type, $format, $context);
+    }
+
+    /**
+     * @param array<int|string,mixed> $context
+     * @return array<int|string,mixed>
+     */
+    public function normalize(mixed $object, string $format = null, array $context = [])
+    {
+        if ($object instanceof Node) {
+            return $this->serializeNode($object);
+        }
+        return $this->fusionContextSerializer->normalize($object, $format, $context);
+    }
+
+    /**
+     * @param array<int|string,mixed> $serializedNode
+     */
+    private function deserializeNode(array $serializedNode): Node
+    {
+        $contentRepositoryId = ContentRepositoryId::fromString($serializedNode['contentRepositoryId']);
+
+        $contentRepository = $this->contentRepositoryRegistry->get($contentRepositoryId);
+
+        $workspace = $contentRepository->getWorkspaceFinder()->findOneByName(WorkspaceName::fromString($serializedNode['workspaceName']));
+        if (!$workspace) {
+            throw new \RuntimeException(sprintf('Could not find workspace while trying to convert node from array %s.', json_encode($serializedNode)), 1699782153);
+        }
+
+        $subgraph = $contentRepository->getContentGraph()->getSubgraph(
+            $workspace->currentContentStreamId,
+            DimensionSpacePoint::fromArray($serializedNode['dimensionSpacePoint']),
+            $workspace->isPublicWorkspace()
+                ? VisibilityConstraints::frontend()
+                : VisibilityConstraints::withoutRestrictions()
+        );
+
+        $node = $subgraph->findNodeById(NodeAggregateId::fromString($serializedNode['nodeAggregateId']));
+        if (!$node) {
+            throw new \RuntimeException(sprintf('Could not find serialized node %s.', json_encode($serializedNode)), 1699782153);
```

Hm, isn't this a probable case that a node was stored which doesn't exist anymore now?
Should we fail hard here, or do we have another way of handling this?

---

Yes we have no choice but to fail.

Alternatively we could just return `null`, which will most likely completely crash silently in the rendering something.

The idea is that if you use `@cache.context = { 0 = 'node' }` (which will write a cache entry into the content cache) and then delete the node, the whole fusion rendering cannot happen in the first place because the node is deleted.
Unless of course caching ... But cache flushing must cover this. Or a hard `cache:flushone Neos_Fusion_Content` 😅 

An interesting test case still ....

In 8.3 we also were not happy if the node could not be converted: https://github.com/neos/neos-development-collection/blob/072186e1601d82a8565942a05c9fe7253cf7f85d/Neos.ContentRepository/Classes/TypeConverter/NodeConverter.php#L166-L169

---

That's not the same, in 8.3 this Error would result in the PropertyMapper returning null, thus mostly just breaking rejndering of the sub part here that needs the node. I think that's saner. You cannot control editor behavior...

---

thanks for the hint ^^ should we log anything in the silent fail case?

---

Aaaand adjusted to return null ;)  If the workspace doesnt exist, i will still throw as this is most likely a fundamental error and not an editor hickup... i think ...